sse2neon.h

```diff
@@ -1264,164 +1264,418 @@ FORCE_INLINE __m128i _mm_shuffle_epi32_default(__m128i a,
+    return vdupq_lane_s16(absolute, 0);
+#endif
+}
+
+FORCE_INLINE int32x4_t _mm_sll_to_vshl_s32(__m128i _count)
+{
+    int64x2_t count = vreinterpretq_s64_m128i(_count);
+#ifdef SSE2NEON_UNSAFE_SHIFTS
+    return vdupq_lane_s32(vget_low_s32(vreinterpretq_s32_s64(count)), 0);
+#else
+    int32x2_t masked_32 = vqmovn_s64(count);
+    int32x2_t absolute = vqabs_s32(masked_32);
+    return vdupq_lane_s32(absolute, 0);
+#endif
+}
+
+FORCE_INLINE int64x2_t _mm_sll_to_vshl_s64(__m128i _count)
+{
+    int64x2_t count = vreinterpretq_s64_m128i(_count);
+#ifndef SSE2NEON_UNSAFE_SHIFTS
+    count = vqabsq_s64(count);
+#endif
+    return vdupq_lane_s64(vget_low_s64(count), 0);
+}
+
+// The immediate forms of vshl/vshr are avoided due to Clang being extremely
+// picky about what it calls a literal, and because the vector form handles
+// larger shifts the same way as SSE2.
+
+// 16-bit shifts
+
+// Shifts the 8 signed or unsigned 16-bit integers left by count bits
+// while shifting in zeroes.
+//
+//   r0 := a0 << count
+//   r1 := a1 << count
+//   r2 := a2 << count
+//   ...
+//   r7 := a7 << count immediate
+FORCE_INLINE __m128i _mm_slli_epi16(__m128i a, int count)
+{
+    return vreinterpretq_m128i_u16(
+        vshlq_u16(
+            vreinterpretq_u16_m128i(a),
+            vdupq_n_s16(count & 0xff)
+        )
+    );
+}
+
+// Shifts the 8 unsigned 16-bit integers right by count bits while shifting
+// in zeroes.
 //
 //   r0 := a0 >> count
 //   r1 := a1 >> count
 //   r2 := a2 >> count
-//   r3 := a3 >> count immediate
-FORCE_INLINE __m128i _mm_srai_epi32(__m128i a, int count)
+//   ...
+//   r7 := a7 >> count immediate
+FORCE_INLINE __m128i _mm_srli_epi16(__m128i a, int count)
 {
-    return (__m128i) vshlq_s32((int32x4_t) a, vdupq_n_s32(-count));
+    return vreinterpretq_m128i_u16(
+        vshlq_u16(
+            vreinterpretq_u16_m128i(a),
+            vdupq_n_s16(-(count & 0xff))
+        )
+    );
 }
 
-// Shifts the 8 signed 16-bit integers in a right by count bits while shifting
+// Shifts the 8 signed 16-bit integers right by count bits while shifting
 // in the sign bit.
 //
 //   r0 := a0 >> count
 //   r1 := a1 >> count
+//   r2 := a2 >> count
 //   ...
-//   r7 := a7 >> count
+//   r7 := a7 >> count immediate
 FORCE_INLINE __m128i _mm_srai_epi16(__m128i a, int count)
 {
-    return (__m128i) vshlq_s16((int16x8_t) a, vdupq_n_s16(-count));
+    return vreinterpretq_m128i_s16(
+        vshlq_s16(
+            vreinterpretq_s16_m128i(a),
+            vdupq_n_s16(-(count & 0xff))
+        )
+    );
+}
+
+// Shift packed 16-bit integers in a left by the amount specified by
+// the corresponding element in count while shifting in zeros, and
+// store the results in dst.
+FORCE_INLINE __m128i _mm_sllv_epi16(__m128i _a, __m128i _count)
+{
+    uint16x8_t a = vreinterpretq_u16_m128i(_a);
+    int16x8_t count = _mm_sllv_to_vshl_s16(_count);
+    return vreinterpretq_m128i_u16(vshlq_u16(a, count));
+}
```

Comment:
I think it still has potential behaviour mismatching when one of the element of `_count` is `0bxxxxxxx1 00000001`.
And **vqabsq_s16(count)** seems not helping in this case?

Reply:
Oh, yeah, if `0 > amt > -16`, that would cause a problem. Silly me.

This should work, right?
```c
FORCE_INLINE __m128i _mm_sllv_epi16(__m128i val, __m128i count)
{
    // count < 0 ? 0xffff : 0x0000
    int16x8_t mask = vshrq_n_s16(count, 15);
    // count > 0 ? (val << count) : (val >> count) 
    uint16x8_t shifted = vshlq_u16(val, count);
    // Mask out any negative shifts
    // shifted & (~mask)
    return vbicq_u16(shifted, mask);
}

FORCE_INLINE __m128i _mm_srlv_epi16(__m128i val, __m128i count)
{
    // count < 0 ? 0xffff : 0x0000
    int16x8_t mask = vshrq_n_s16(count, 15);
    // negative = shift right
    int16x8_t count_neg = vnegq_s16(count);
    // count_neg > 0 ? (val << count_neg) : (val >> count_neg) 
    uint16x8_t shifted = vshlq_u16(val, count_neg);
    // Mask out any negative shifts
    // shifted & (~mask)
    return vbicq_u16(shifted, mask);
}


FORCE_INLINE __m128i _mm_srav_epi16(__m128i val, __m128i count)
{
    // count < 0 ? 0xffff : 0x0000
    int16x8_t mask = vshrq_n_s16(count, 15);
    // negative = shift right
    int16x8_t countNeg = vnegq_s16(count);
    // count_neg > 0 ? (val << count_neg) : (val >> count_neg) 
    int16x8_t shifted = vshlq_s16(val, count_neg);
    // For large shifts, just extend the sign bit
    int16x8_t sign_mask = vshrq_n_s16(val, 15);
    // Either use the sign mask or the actual value, depending
    // on whether the shift was greater than zero
    return vbslq_s16(mask, sign_mask, shifted);
}
```

Reply:
And for the single ones, this seems to be pretty good:
```c
__m128i _mm_sll_epi16(__m128i val, __m128i count)
{
    // get the important part of count
    int16x8_t shifter = vdupq_lane_s16(vreinterpret_s16_u64(vget_low_u64(count)), 0);
    // Shift left with the shifter
    uint16x8_t shift = vshlq_u16(val, shifter);                                                      
    // Check if count[0] > 15
    // We use a shift to avoid moving immediates from scalar.
#ifdef __aarch64__
    // (count >> 4) == 0
    uint64x1_t shifted = vshr_n_u64(vget_low_u64(count), 4);
    uint16x8_t mask = vreinterpretq_u16_u64(vdupq_lane_u64(vceqz_u64(shifted), 0));
#else
    // no compare for uint64x1_t, improvise
    // narrow = saturate_u32(count >> 4)
    // any large values will saturate to 0xffffffff
    uint32x2_t narrow = vqshrn_n_u64(count, 4);
    // check if zero
    uint16x8_t mask = vreinterpretq_u16_u32(vdupq_lane_u32(vceq_u32(narrow, vdup_n_u32(0)), 0));
#endif
    // AND out the garbage
    return vandq_u16(shift, mask);
}
```

Reply:
Take **_mm_sllv_epi16** as an example.

```c
    // count < 0 ? 0xffff : 0x0000
    int16x8_t mask = vshrq_n_s16(count, 15);
```

I think the result is `// count < 0 ? 0x0001 : 0x0000` since it shift right 15 bits?

```c
    // count > 0 ? (val << count) : (val >> count) 
    uint16x8_t shifted = vshlq_u16(val, count);
    // Mask out any negative shifts
    // shifted & (~mask)
    return vbicq_u16(shifted, mask);
```

The mask is not reversed in `return vbicq_u16(shifted, mask)`.

It still not handles the situation I mentioned before.
For example, what if the shift value is `0x0101` = `257`?
According to [_mm_sllv_epi16](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_sllv_epi16&expand=5326), it would shift left `257` bits.
However, the **vshlq_u16** would shift left only `1` bit since
```
This instruction takes each element in the vector of the first source SIMD&FP register,
shifts each element by a value from the least significant byte
of the corresponding element of the second source SIMD&FP register,
```

Reply:
It is a *signed* shift right. I'm just going to do vclt instead (aside from s64 because v7).

Reply:
You are right, I did not notice that it is `signed shift right`.

Reply:
Oh, I see what you mean by the 257 thing…

It reads an int8_t, not the entire integer.

Time for yet another iteration.

Reply:
Okay, I think I can just do this
```c
FORCE_INLINE __m128i _mm_sllv_epi32(__m128i a, __m128i amt)
{
      uint32x4_t mask = vcltq_u32((uint32x4_t)amt, vdupq_n_u32(32));
      uint32x4_t shifted = vshlq_u32((uint32x4_t)a, (uint32x4_t)amt);
      return (__m128i)vandq_u32(shifted, mask);
}
```

Reply:
It seems work well.
Please modify the related intrinsic functions as well.